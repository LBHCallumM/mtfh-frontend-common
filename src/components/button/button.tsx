import React, { ElementType, forwardRef } from 'react';
import cn from 'classnames';

import { PolymorphicComponentProps } from '../../types';
import './styles.scss';

export interface ButtonOwnProps {
  variant?: 'primary' | 'secondary';
  isLoading?: boolean;
  isDisabled?: boolean;
  loadingText?: string;
}

export type ButtonProps<C extends ElementType> = PolymorphicComponentProps<
  C,
  ButtonOwnProps
>;

function ButtonWithRef<C extends ElementType = 'button'>(
  {
    as,
    variant = 'primary',
    isLoading = false,
    loadingText,
    isDisabled,
    children,
    className,
    ...props
  }: ButtonProps<C>,
  ref: any
) {
  const buttonClasses = cn('govuk-button', 'lbh-button', {
    'govuk-secondary lbh-button--secondary': variant === 'secondary',
    'lbh-button--disabled govuk-button--disabled': isDisabled,
    className,
  });

  const Component = as || 'button';

  return (
    <Component
      ref={ref}
      className={buttonClasses}
      disabled={isDisabled}
      aria-disabled={isDisabled}
      {...props}
    >
      {isLoading && (
        <span className="button-loading-indicator">
          <span>Loading...</span>
        </span>
      )}
      {isLoading && loadingText ? loadingText : children}
    </Component>
  );
}

export const Button: <C extends ElementType = 'button'>(
  props: ButtonProps<C>
) => JSX.Element | null = forwardRef(ButtonWithRef);
