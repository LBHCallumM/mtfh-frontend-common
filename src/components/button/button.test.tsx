import React from 'react';
import { render, screen } from '@testing-library/react';

import { Button } from './button';

test('it renders correctly', () => {
  const { container } = render(<Button>Test Button</Button>);
  expect(container).toMatchSnapshot();
});

test('it shows a loading state', () => {
  const { container } = render(<Button isLoading>Test Button</Button>);
  expect(container).toMatchSnapshot();
});

test('it shows a loading state with custom text', () => {
  render(
    <Button isLoading loadingText="Processing">
      Test Button
    </Button>
  );
  expect(screen.getByText('Processing')).toBeInTheDocument();
});

test('it can support anchor tag', () => {
  const { container } = render(
    <Button as="a" href="http://localhost">
      Test Anchor
    </Button>
  );
  expect(container).toMatchSnapshot();
});
