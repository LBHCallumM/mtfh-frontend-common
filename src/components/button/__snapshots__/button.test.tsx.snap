// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it can support anchor tag 1`] = `
<div>
  <a
    class="govuk-button lbh-button"
    href="http://localhost"
  >
    Test Anchor
  </a>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <button
    class="govuk-button lbh-button"
  >
    Test Button
  </button>
</div>
`;

exports[`it shows a loading state 1`] = `
<div>
  <button
    class="govuk-button lbh-button"
  >
    <span
      class="button-loading-indicator"
    >
      <span>
        Loading...
      </span>
    </span>
    Test Button
  </button>
</div>
`;
